src/transaction.rs

```diff
@@ -1,42 +1,127 @@
-        client: SignerMiddleware<M, S>,
-    ) -> anyhow::Result<Self> {
-        Ok(Self {
-            transaction_request: transaction_request.to_eip1559(),
-            client,
-        })
+impl<C: SolCall> WriteContractParameters<C> {
+    pub fn default() -> Self {
+        Self {
+            call: None,
+            address: None,
+            gas: None,
+            gas_price: None,
+            max_fee_per_gas: None,
+            max_priority_fee_per_gas: None,
+            nonce: None,
+            value: None,
+        }
+    }
+
+    /// Sets the `call` field in the transaction to the provided value
+    /// The call must implement the `SolCall` trait
+    pub fn with_call(mut self, call: C) -> Self {
+        self.call = Some(call);
+        self
+    }
+
+    /// Sets the `address` field in the transaction to the provided value
+    pub fn with_address<T: Into<Address>>(mut self, address: T) -> Self {
+        self.address = Some(address.into());
+        self
+    }
+
+    /// Sets the `gas` field in the transaction to the provided value
+    pub fn with_gas<T: Into<U256>>(mut self, gas: Option<T>) -> Self {
+        self.gas = gas.map(|val| val.into());
+        self
+    }
+
+    /// Sets the `gas_price` field in the transaction to the provided value
+    pub fn with_gas_price<T: Into<U256>>(mut self, gas_price: Option<T>) -> Self {
+        self.gas_price = gas_price.map(|val| val.into());
+        self
     }
 
-    pub async fn from_ethers_transaction_request(
-        transaction_request: Eip1559TransactionRequest,
-        client: SignerMiddleware<M, S>,
-    ) -> anyhow::Result<Self> {
-        Ok(Self {
-            transaction_request,
-            client,
-        })
+    /// Sets the `max_priority_fee_per_gas` field in the transaction to the provided value
+    pub fn with_max_priority_fee_per_gas<T: Into<U256>>(
+        mut self,
+        max_priority_fee_per_gas: Option<T>,
+    ) -> Self {
+        self.max_priority_fee_per_gas = max_priority_fee_per_gas.map(|val| val.into());
+        self
     }
 
-    // Execute the transaction
-    pub async fn execute(&self) -> anyhow::Result<TransactionReceipt> {
+    /// Sets the `max_fee_per_gas` field in the transaction to the provided value
+    pub fn with_max_fee_per_gas<T: Into<U256>>(mut self, max_fee_per_gas: Option<T>) -> Self {
+        self.max_fee_per_gas = max_fee_per_gas.map(|val| val.into());
+        self
+    }
+
+    /// Sets the `nonce` field in the transaction to the provided value
+    pub fn with_nonce<T: Into<U256>>(mut self, nonce: Option<T>) -> Self {
+        self.nonce = nonce.map(|val| val.into());
+        self
+    }
+
+    /// Sets the `value` field in the transaction to the provided value
+    pub fn with_value<T: Into<U256>>(mut self, value: Option<T>) -> Self {
+        self.value = value.map(|val| val.into());
+        self
+    }
+}
+
+pub struct WriteContract<M: Middleware, S: Signer> {
```

Comment:
I might name this something like `WriteClient` or `WritableClient` -- having `Contract` in the name implies to me that its storing the contract address in the struct

Reply:
Also since there's only one struct property, I might make it a bare wrapper to the SignerMiddleware: i.e. `pub struct WriteClient<M: Middleware, S: Signer>(SignerMiddleware<M, S>);`

Reply:
done these